src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -718,6 +812,47 @@ ConsumeQueryResult(PGconn *connection, int64 *rows)
 
 
 /*
+ * ExecuteTransactionEnd ends any remote transactions still taking place on
+ * remote nodes. It uses txnPlacementList to know which placements still need
+ * final COMMIT or ABORT commands. Any failures are added to the other list,
+ * failedTxnPlacementList, to eventually be marked as failed.
+ */
+static void
+ExecuteTransactionEnd(bool commit)
+{
+	char *sqlCommand = commit ? "COMMIT TRANSACTION" : "ABORT TRANSACTION";
+	HASH_SEQ_STATUS scan;
+	XactParticipantEntry *participant;
+
+	hash_seq_init(&scan, xactParticipantHash);
+	while ((participant = (XactParticipantEntry *) hash_seq_search(&scan)))
+	{
+		PGconn *connection = participant->connection;
+		PGresult *result = NULL;
+
+		if (PQstatus(connection) != CONNECTION_OK)
+		{
+			continue;
+		}
+
+		result = PQexec(connection, sqlCommand);
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			WarnRemoteError(connection, result);
+			PurgeConnection(connection);
+
+			if (commit)
+			{
+				MarkParticipantShardsUnhealthy(participant);
```

Comment:
Fixed this. Basically, if no node `COMMIT`s, we'll throw an error, rolling back any health modifications.